README.md
# winston-rsyslog2

A [Rsyslog][0] transport for [winston][1].

## Usage
``` js
  var winston = require('winston');

  //
  // Requiring `winston-rsyslog` will expose 
  // `winston.transports.Rsyslog`
  //
  require('winston-rsyslog2');

  winston.add(winston.transports.Rsyslog, options);
```

*This is a fork of `winston-rsyslog` version 0.0.2. See [the CHANGELOG](./CHANGELOG.md) for differences.*

The rsyslog transport takes the following options:

* __level:__ Level of messages that this transport should log
* __host:__ Host where rsyslog runs (default : localhost)
* __port:__ rsyslog Port (default : 514)
* __facility:__ Facility index (default is 0, valid values are from 0 to 23)
* __protocol:__ TCP or UDP (values can be "U" or "T", default is "U")
* __hostname:__ The hostname the application is running at (default is the current hostname)
* __tag:__ A tag to name the application for easy log filtering (default is 'winston')
* __timeout:__ Socket timeout for TCP (default is 2000 ms)
* __levelMapping:__ Mapping of level names to rsyslog security levels (defaults to `winston.config.syslog.levelMapping`)
* __dateProvider:__ A parameterless function that returns a date string (default is `new Date().toISOString()`)
* __messageProvider:__ A function that returns the message string, takes three parameters: `level`, `msg` and `meta` (`msg` is already formatted by winston)

Through __facility__ parameter, You'll be able to use the right log file on rsyslog:

```
0 :  kernel messages
1 :  user-level messages
2 :  mail system
3 :  system daemons
4 :  security/authorization messages
5 :  messages generated internally by syslogd
6 :  line printer subsystem
7 :  network news subsystem
8 :  UUCP subsystem
9 :  clock daemon
10 : security/authorization messages
11 : FTP daemon
12 : NTP subsystem
13 : log audit
14 : log alert
15 : clock daemon (note 2)
16 : local use 0 (local0)
17 : local use 1 (local1)
18 : local use 2 (local2)
19 : local use 3 (local3)
20 : local use 4 (local4)
21 : local use 5 (local5)
22 : local use 6 (local6)
23 : local use 7 (local7)
```

## Enabling rsyslog
To let rsyslog receive data through UDP port, you should change its configuration (typically located in /etc/rsyslog.conf),
enabling the following tags:

``` bash
$ModLoad imudp
$UDPServerRun 514
```

The first tag enables UDP reception, while the second one defines the listening port.

To enable TCP reception on port 10514 (as an example):

``` bash
$ModLoad imtcp
$InputTCPServerRun 10514
```

## Installation

``` bash
npm install winston-rsyslog2
```

## Run Tests
All of the winston tests are written in [vows][3], and cover all of the use cases described above. 
To configure the destination rsyslog daemon parameters, copy the test/config.example.json file into test/config.json 
and modify it to suite your needs.
If no test/config.json file is found, defaults will be used.

Then you can run tests with [npm][4]:

``` bash
npm test
```

#### Author: [Fabio Grande](http://about.me/fgrande)
#### License: MIT

[0]: http://en.wikipedia.org/wiki/Rsyslog
[1]: https://github.com/flatiron/winston
[2]: https://github.com/indexzero/winston-loggly
[3]: http://vowsjs.org
[4]: http://npmjs.org


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.0.5] - 2015-05-22

### Changed

- Improved README and Changelog formatting. (Mark Stosberg)

### Fixed

- Repaired testing function broken in 0.0.4. (Mark Stosberg)

## [0.0.4] - 2015-05-22

### Fixed

- license is now clarified in package.json

### Added

 - `timeout` option for TCP timeout. Default value is 2000 ms, which was the prior hardcoded behavior. (Aleksey verkholantsev)
 - `dateProvider` option to generate a different date string. Default behavior is backwards compatible.  (Aleksey verkholantsev)
 - `messageProvider` option to customize the message string forfmatting. (Aleksey verkholantsev)
 - `levelMapping` option to map winston logging levels to rsyslog

## [0.0.2] - 2015-05-20

### Changed

- First release as fork from winston-rsyslog2.

### Fixed

- Dependency on winston was changed to be ">=" so it can be used with newer versions of wiston.


test/winston-rsyslog-test.js
/*
 * winston-rsyslog-test.js: Tests for instances of the Rsyslog transport
 *
 * (C) 2013-2015 Fabio Grande
 * MIT LICENSE
 *
 */

var path = require('path'),
  vows = require('vows'),
  assert = require('assert'),
  winston = require('winston'),
  helpers = require('winston/test/helpers'),
  Rsyslog = require('../lib/winston-rsyslog').Rsyslog;

var tokenTransport, config;

try {
  config = require('./config');
  console.log('Using configuration file test/config.json.');
  tokenTransport = new Rsyslog(config.transports.rsyslog);
} catch (e) {
  console.log('Cannot read file test/config.json. Using defaults.');
  tokenTransport = new Rsyslog();
}

function assertRsyslog(transport) {
  assert.instanceOf(transport, Rsyslog);
  assert.isFunction(transport.log);
}

vows.describe('winston-rsyslog').addBatch({
    'An instance of the Rsyslog Transport': {
        'should have the proper methods defined': function() {
            assertRsyslog(tokenTransport);
        },
        'the log() method': helpers.testSyslogLevels(tokenTransport, 'should log messages to rsyslog', function(ign, err, logged) {
            assert.isNull(err);
            assert.isTrue(logged);
        })
    }
}).export(module);


lib/winston-rsyslog.js
/*
 * winston-rsyslog: Transport for logging to remote syslog
 *
 * (C) 2013-2015 Fabio Grande
 * MIT LICENSE
 */

var events = require('events'),
  dgram = require('dgram'),
  net = require('net'),
  os = require('os'),
  util = require('util'),
  winston = require('winston'),
  Transport = winston.Transport;

/**
 * Constructor function for the rsyslog transport object responsible
 * for sending messages to syslog daemon
 * @param {Object} [options] Options for this instance
 * @constructor
 */
var Rsyslog = exports.Rsyslog = function (options) {
  options = options || {};
  Transport.call(this, options);

  this.name = 'rsyslog';
  this.host = options.host || 'localhost';
  this.port = options.port || 514;
  this.facility = options.facility || 0;
  this.protocol = options.protocol || 'U';
  this.hostname = options.hostname || os.hostname();
  this.tag = options.tag || 'winston';
  this.timeout = options.timeout || 2000;
  this.levelMapping = options.levelMapping || {};
  this.dateProvider = options.dateProvider || dateProviderDefault;
  this.messageProvider = options.messageProvider || messageProviderDefault;

  if (this.facility > 23 || this.facility < 0) {
    throw new Error('Facility index is out of range! (valid range is 0..23)');
  }

  if (this.protocol != 'U' && this.protocol != 'T') {
    throw new Error('Undefined Protocol! (valid options are U or T)');
  }
};

// Inherit from `winston.Transport`.
util.inherits(Rsyslog, winston.Transport);

// Add a new property to expose the new transport....
winston.transports.Rsyslog = Rsyslog;

// Expose the name of this Transport on the prototype
Rsyslog.prototype.name = 'rsyslog';

/**
 * Core logging method exposed to Winston. Metadata is optional.
 * @param {string} level Level at which to log the message
 * @param {string} msg Message to log
 * @param {Object} [meta] Additional metadata to attach
 * @callback callback Called on completion
 */
Rsyslog.prototype.log = function (level, msg, meta, callback) {
  if (this.silent) {
    return callback(null, true);
  }

  // If the specified level is not included in the syslog list, convert it to 'debug'
  var _severity = 7;
  if (this.levelMapping[level] !== undefined) {
    _severity = this.levelMapping[level];
  }

  var _pri = (this.facility << 3) + _severity;
  var _date = this.dateProvider();
  var _message = this.messageProvider(level, msg, meta);
  var _buffer = new Buffer('<' + _pri + '>' + _date + ' ' + this.hostname + ' ' + this.tag + ' ' + _message);

  if (this.protocol === 'U') {
    sendUdp(this, _buffer, callback);
  } else if (this.protocol === 'T') {
    sendTcp(this, _buffer, callback);
  }
};

function sendUdp(self, buffer, callback) {
  var client = dgram.createSocket('udp4');
  client.send(buffer, 0, buffer.length, self.port, self.host, function (err) {
    if (err) {
      if (callback) {
        return callback(err);
      }
      throw err;
    }

    self.emit('logged');

    if (callback) {
      callback(null, true);
    }

    client.close();
  });
}

function sendTcp(self, buffer, callback) {
  var socket = net.connect(self.port, self.host, function () {
    socket.end(buffer + '\n');

    self.emit('logged');

    if (callback) {
      callback(null, true);
    }
  });

  socket.setTimeout(self.timeout);

  socket.on('error', function (err) {
    socket.destroy();
    if (callback) {
      return callback(err);
    }

    throw err;
  });

  socket.on('timeout', function (err) {
    socket.destroy();
    if (callback) {
      return callback(err);
    }

    throw err;
  });
  return callback;
}

var dateProviderDefault = function () {
  return new Date().toISOString();
};

var messageProviderDefault = function (level, msg, meta) {
  var message = process.pid + ' - ' + level + ' - ' + msg;
  if (meta && (typeof meta !== 'object' || Object.keys(meta).length)) {
    message += ' ' + util.inspect(meta);
  }
  return message;
};
